core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1935,4 +2067,107 @@ object Log {
+      val overflowOffset = overflowBatches.headOption.map(firstBatch => {
+        log.info(s"Found overflow at offset ${firstBatch.baseOffset} in segment $segment of log $log")
+        firstBatch.baseOffset
+      })
+
+      // return early if no valid batches were found
+      if (validBatches.isEmpty) {
+        require(overflowOffset.isDefined, "No batches found during split")
+        return new CopyResult(0, overflowOffset)
+      }
+
+      // determine the maximum offset and timestamp in batches
+      for (batch <- validBatches) {
+        if (batch.maxTimestamp > maxTimestamp) {
+          maxTimestamp = batch.maxTimestamp
+          offsetOfMaxTimestamp = batch.lastOffset
+        }
+        maxOffset = batch.lastOffset
+        bytesRead += batch.sizeInBytes
+      }
+
+      // read all valid batches into memory
+      val validRecords = records.slice(0, bytesRead)
+      require(readBuffer.capacity >= validRecords.sizeInBytes)
+      readBuffer.clear()
+      readBuffer.limit(validRecords.sizeInBytes)
+      validRecords.readInto(readBuffer, 0)
+
+      // append valid batches into the segment
+      segment.append(maxOffset, maxTimestamp, offsetOfMaxTimestamp, MemoryRecords.readableRecords(readBuffer))
+      readBuffer.clear()
+      log.info(s"Appended messages till $maxOffset to segment $segment during split")
+
+      new CopyResult(bytesRead, overflowOffset)
+    }
+
+    log.info(s"Splitting segment $segment in log $log")
+    newSegments += LogCleaner.createNewCleanedSegment(log, segment.baseOffset)
+    while (position < sourceRecords.sizeInBytes) {
+      val currentSegment = newSegments.last
+
+      // grow buffers if needed
+      val firstBatch = sourceRecords.batchesFrom(position).asScala.head
+      if (firstBatch.sizeInBytes > readBuffer.capacity)
+        readBuffer = ByteBuffer.allocate(firstBatch.sizeInBytes)
+
+      // get records we want to copy and copy them into the new segment
+      val recordsToCopy = sourceRecords.slice(position, readBuffer.capacity)
+      val copyResult = copyRecordsToSegment(recordsToCopy, currentSegment, readBuffer)
+      position += copyResult.bytesRead
+
+      // create a new segment if there was an overflow
+      copyResult.overflowOffset.foreach(overflowOffset => newSegments += LogCleaner.createNewCleanedSegment(log, overflowOffset))
+    }
+    require(newSegments.length > 1, s"No offset overflow found for $segment in $log")
+
+    // prepare new segments
+    log.info(s"Split messages from $segment of $log into ${newSegments.length} new segments")
+    newSegments.foreach { splitSegment =>
+      splitSegment.onBecomeInactiveSegment()
+      splitSegment.flush()
+      splitSegment.lastModified = segment.lastModified
+      log.info(s"New segment: $splitSegment")
+    }
+
+    // replace old segment with new ones
+    log.replaceSegments(newSegments.toList, List(segment), isRecoveredSwapFile = false)
```

Comment:
This logic will get invoked during log initialization if we encounter an overflowed segment. Is that expected? It seems like it will result in the new segments being added to the `segments` collection. But then they would be overwritten in the retry of `loadSegments`.

Reply:
I think we have to call `replaceSegments` so that the old segment can be deleted. I don't think it is worth having a special case only for initial segment load.

Because `loadSegmentFiles` is somewhat of a special case, I moved the exception handling logic that was currently in `loadSegmentFiles()` into `loadSegments()` so that we close all segments after the split.